### ISSUE TITLE
parse_mag_decl cause clippy crash

### ISSUE THREAD

**Issue by Dushistov** [rust-lang-nursery/rust-clippy#2499]
I run "cargo +nightly clippy" for "cargo new --lib foo" crate with such `lib.rs`:

```rust
use std::str;

type ParseLineError = i32;

fn parse_num<T: str::FromStr>(s: &[u8], column: usize, len: usize) -> Result<T, ParseLineError> {
    str::from_utf8(&s[column..(column + len)])
        .map_err(|_| -1)?
        .parse::<T>()
        .map_err(|_| -1)
}

fn parse_mag_decl(s: &[u8], mut column: usize) -> Result<f64, ParseLineError> {
    let t = s[column];
    if !t.is_ascii() {
        return Err(-1);
    }
    column += 1;
    let t = t as char;
    match t {
        'E' | 'W' => {}
        'T' => if &s[column..(column + 4)] != &['0' as u8; 4] {
            return Err(-1);
        } else {
            return Ok(0.);
        },
        _ => {
            return Err(-1);
        }
    }
    let deg = parse_num::<f64>(s, column, 3)?;
    column += 3;
    let deg_tenth = parse_num::<f64>(s, column, 1)?;
    let mut val = (deg + deg_tenth / 10.).to_radians();
    if t == 'W' {
        val = -val;
    }
    Ok(val)
}
```

it crash in such way:

```
error: internal compiler error: librustc/ty/context.rs:478: node_id_to_type: no type for node `expr 4 (id=111)`

thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:535:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:59
             at libstd/panicking.rs:207
   3: std::panicking::default_hook
             at libstd/panicking.rs:223
   4: core::ops::function::Fn::call
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:403
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::session::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: <std::thread::local::LocalKey<T>>::try_with
  11: <std::thread::local::LocalKey<T>>::with
  12: rustc::ty::context::tls::with
  13: rustc::ty::context::tls::with_opt
  14: rustc::session::opt_span_bug_fmt
  15: rustc::session::bug_fmt
  16: rustc::ty::context::TypeckTables::node_id_to_type
  17: rustc::ty::context::TypeckTables::expr_ty
  18: clippy_lints::consts::ConstEvalLateContext::expr
  19: clippy_lints::consts::constant_simple
  20: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
  21: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
  22: <clippy_lints::copies::CopyAndPaste as rustc::lint::LateLintPass<'a, 'tcx>>::check_expr
  23: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
  24: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_stmt
  25: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_block
  26: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
  27: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_body
  28: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_nested_body
  29: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_fn
  30: rustc::hir::intravisit::walk_item
  31: rustc::lint::context::LateContext::with_param_env
  32: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  33: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_mod
  34: rustc::lint::context::check_crate
  35: rustc::ty::context::TyCtxt::create_and_enter
  36: rustc_driver::driver::compile_input
  37: rustc_driver::run_compiler
```

```sh
$ cargo-clippy --version
0.0.187
```

cargo +nightly build works fine for this crate without any crash.

**Comment by Dushistov** [comment]
Reduced test case:

```rust
fn f(s: &[u8]) -> bool {
    let t = s[0] as char;

    match t {
        'E' | 'W' => {}
        'T' => if &s[0..(0 + 4)] != &['0' as u8; 4] {
            return false;
        } else {
            return true;
        },
        _ => {
            return false;
        }
    }
    true
}
```

**Comment by Aaron1011** [comment]
Here's the backtrace from debug mode (includes line numbers, functions aren't inlined):

```
error: internal compiler error: librustc/ty/context.rs:478: node_id_to_type: no type for node `expr 4 (id=40)`

thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:535:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:59
             at libstd/panicking.rs:207
   3: std::panicking::default_hook
             at libstd/panicking.rs:223
   4: core::ops::function::Fn::call
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:403
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::session::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: <std::thread::local::LocalKey<T>>::try_with
  11: <std::thread::local::LocalKey<T>>::with
  12: rustc::ty::context::tls::with
  13: rustc::ty::context::tls::with_opt
  14: rustc::session::opt_span_bug_fmt
  15: rustc::session::bug_fmt
  16: rustc::ty::context::TypeckTables::node_id_to_type
  17: rustc::ty::context::TypeckTables::expr_ty
  18: clippy_lints::consts::ConstEvalLateContext::expr
             at clippy_lints/src/consts.rs:258
  19: clippy_lints::consts::constant
             at clippy_lints/src/consts.rs:225
  20: clippy_lints::consts::constant_simple
             at clippy_lints/src/consts.rs:229
  21: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
             at clippy_lints/src/utils/hir_utils.rs:320
  22: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
             at clippy_lints/src/utils/hir_utils.rs:464
  23: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
             at clippy_lints/src/utils/hir_utils.rs:329
  24: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
             at clippy_lints/src/utils/hir_utils.rs:366
  25: clippy_lints::utils::hir_utils::SpanlessHash::hash_expr
             at clippy_lints/src/utils/hir_utils.rs:420
  26: clippy_lints::copies::lint_match_arms::{{closure}}
             at clippy_lints/src/copies.rs:178
  27: clippy_lints::copies::search_same
             at clippy_lints/src/copies.rs:338
  28: clippy_lints::copies::lint_match_arms
             at clippy_lints/src/copies.rs:194
  29: <clippy_lints::copies::CopyAndPaste as rustc::lint::LateLintPass<'a, 'tcx>>::check_expr
             at clippy_lints/src/copies.rs:130
  30: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
  31: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_stmt
  32: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_block
  33: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
  34: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_body
  35: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_nested_body
  36: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_fn
  37: rustc::hir::intravisit::walk_item
  38: rustc::lint::context::LateContext::with_param_env
  39: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  40: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_mod
  41: rustc::lint::context::check_crate
  42: rustc::ty::context::TyCtxt::create_and_enter
  43: rustc_driver::driver::compile_input
  44: rustc_driver::run_compiler
  45: clippy_driver::main::{{closure}}
             at src/driver.rs:197
  46: rustc_driver::run::{{closure}}
             at /checkout/src/librustc_driver/lib.rs:172
  47: rustc_driver::monitor::{{closure}}
             at /checkout/src/librustc_driver/lib.rs:1505
```

**Comment by oli-obk** [comment]
Ah I already have a fix for that one in the miri PR on this repo

**Comment by phansch** [comment]
This will probably be fixed together with #2594
